### ISSUE TITLE
MosaicML LLM-Foundry Quick Start in Paperspace with Non-Ampere GPU (using `vishalbakshi/llm-foundry-jupyter:latest` image)

### ISSUE THREAD

**Issue by vishalbakshi** [LocalResearchGroup/llm-foundry#2]
Documenting steps to run the llm-foundry quick start in a Paperspace Pro Gradient Notebook with a P5000 GPU.

## Creating a Notebook

Select "Start from Scratch" as the template to use.

Click "Advanced Options" and put `vishalbakshi/llm-foundry-jupyter:latest` in Container > Name

Click "Start notebook"

Click start (note: I got an error about workspace username being empty so I had to do these steps twice to get it to run.) It will take a few minutes for the image to be built.

## Quick Start Modified Script

Note that the script below is not exactly the same as the llm-foundry Quick Start as I have modified the following script (with Claude's help) for a P5000 16GB GPU which is not an Ampere GPU (and was running into OOM errors with the original script). Feel free to experiment with this. Create a new file, I called it `train_mpt.sh` and copy paste the following code into it:

```
#!/bin/bash
cd /llm-foundry/scripts

# Convert C4 dataset to StreamingDataset format
# Note: concat_tokens=1024 to resolve OOM error on 16GB P5000 GPU
python data_prep/convert_dataset_hf.py \
  --dataset allenai/c4 --data_subset en \
  --out_root my-copy-c4 --splits train_small val_small \
  --concat_tokens 1024 --tokenizer EleutherAI/gpt-neox-20b --eos_text '<|endoftext|>'

# Train an MPT-125m model for 10 batches
# Note: Using torch attention and fp32 for Non-Ampere GPUs
# Note: Reduced batch sizes and seq_len for 16GB P5000 GPU
composer train/train.py \
  train/yamls/pretrain/mpt-125m.yaml \
  variables.data_local=my-copy-c4 \
  variables.max_seq_len=1024 \
  train_loader.dataset.split=train_small \
  eval_loader.dataset.split=val_small \
  max_duration=10ba \
  eval_interval=0 \
  save_folder=mpt-125m \
  model.attn_config.attn_impl=torch \
  device_train_microbatch_size=4 \
  global_train_batch_size=32 \
  precision=fp32

# Convert the model to HuggingFace format
# Note: Using fp32 for Non-Ampere GPUs
python inference/convert_composer_to_hf.py \
  --composer_path mpt-125m/ep0-ba10-rank0.pt \
  --hf_output_path mpt-125m-hf \
  --output_precision fp32

# Evaluate the model on a subset of tasks
# Note: Using fp32 for Non-Ampere GPUs
composer eval/eval.py \
  eval/yamls/hf_eval.yaml \
  icl_tasks=eval/yamls/copa.yaml \
  variables.model_name_or_path=mpt-125m-hf \
  precision=fp32

# Generate responses to prompts
python inference/hf_generate.py \
  --name_or_path mpt-125m-hf \
  --max_new_tokens 256 \
  --prompts \
    "The answer to life, the universe, and happiness is" \
    "Here's a quick recipe for baking chocolate chip cookies: Start by"

```

Run it in the Paperspace terminal with:

```
/notebooks/train_mpt.sh
```

## Dataset prep, training and evaluation logs

I have copy pasted some of the logs below for reference. 

```
# /notebooks/train_mpt.sh
Resolving data files: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1024/1024 [00:05<00:00, 194.71it/s]
Resolving data files: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1024/1024 [00:00<00:00, 30834.94it/s]
Converting train_small to MDS format...
Note: the progress bar is based on the dataset length before tokenization, and may finish at a value before 100%.
train_small: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 100000/100000 [02:13<00:00, 750.26it/s]
Resolving data files: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1024/1024 [00:04<00:00, 248.99it/s]
Resolving data files: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1024/1024 [00:00<00:00, 2773.69it/s]
Converting val_small to MDS format...
Note: the progress bar is based on the dataset length before tokenization, and may finish at a value before 100%.
val_small: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10000/10000 [00:21<00:00, 464.73it/s]

2025-01-26 16:24:55,458: rank0[4251][MainThread]: DEBUG: llmfoundry.models.mpt.modeling_mpt: MPTModel(
  (wte): SharedEmbedding(50368, 768)
  (wpe): Embedding(1024, 768)
  (emb_drop): Dropout(p=0.0, inplace=False)
  (blocks): ModuleList(
    (0-11): 12 x MPTBlock(
      (norm_1): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
      (attn): MultiheadAttention(
        (Wqkv): Linear(in_features=768, out_features=2304, bias=True)
        (out_proj): Linear(in_features=768, out_features=768, bias=True)
      )
      (norm_2): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
      (ffn): MPTMLP(
        (up_proj): Linear(in_features=768, out_features=3072, bias=True)
        (down_proj): Linear(in_features=3072, out_features=768, bias=True)
      )
      (resid_attn_dropout): Dropout(p=0.0, inplace=False)
      (resid_ffn_dropout): Dropout(p=0.0, inplace=False)
    )
  )
  (norm_f): LPLayerNorm((768,), eps=1e-05, elementwise_affine=True)
)

******************************
2025-01-26 16:24:57,549: rank0[4251][MainThread]: DEBUG: composer.trainer.trainer: Spinning the dataloaders
[batch=1/10]:
         Train time/epoch: 0
         Train time/batch: 0
         Train time/sample: 0
         Train time/batch_in_epoch: 0
         Train time/sample_in_epoch: 0
         Train time/token: 0
         Train time/token_in_epoch: 0
         Train memory/current_allocated_mem: 1.8500
         Train memory/current_active_mem: 1.8500
         Train memory/current_inactive_mem: 1.1448
         Train memory/current_reserved_mem: 8.1244
         Train memory/peak_allocated_mem: 7.7141
         Train memory/peak_active_mem: 7.7141
         Train memory/peak_inactive_mem: 1.8273
         Train memory/peak_reserved_mem: 8.1244
         Train memory/alloc_retries: 0
         Train trainer/device_train_microbatch_size: 4
         Train loss/train/total: 11.9162
         Train metrics/train/LanguageCrossEntropy: 11.9162
         Train metrics/train/LanguagePerplexity: 149668.1562
         Train metrics/train/TokenAccuracy: 0.0000
         Train time/train: 0.0030
         Train time/val: 0.0000
         Train time/total: 0.0030
         Train lr-DecoupledAdamW/group0: 0.0000

##############################
Converting Composer checkpoint to HuggingFace checkpoint format...
Downloading checkpoint from mpt-125m/ep0-ba10-rank0.pt -> /tmp/tmpcsx6x94a/local-composer-checkpoint.pt
##############################
Saving HF Model Weights...
##############################
HF checkpoint folder successfully created at mpt-125m-hf.
Loading model from mpt-125m-hf

[Eval batch=1/50] Eval on copa/0-shot data
[Eval batch=6/50] Eval on copa/0-shot data
[Eval batch=11/50] Eval on copa/0-shot data
[Eval batch=16/50] Eval on copa/0-shot data
[Eval batch=21/50] Eval on copa/0-shot data
[Eval batch=26/50] Eval on copa/0-shot data
[Eval batch=30/50] Eval on copa/0-shot data
[Eval batch=35/50] Eval on copa/0-shot data
[Eval batch=40/50] Eval on copa/0-shot data
[Eval batch=45/50] Eval on copa/0-shot data
[Eval batch=50/50] Eval on copa/0-shot data:
         Eval metrics/copa/0-shot/InContextLearningMultipleChoiceAccuracy: 0.5400

Printing gauntlet results for all models
| model_name   |   core_average |   world_knowledge |   commonsense_reasoning |   language_understanding |   symbolic_problem_solving |   reading_comprehension |
|:-------------|---------------:|------------------:|------------------------:|-------------------------:|---------------------------:|------------------------:|
| mpt-125m-hf  |              0 |               nan |                     nan |                      nan |                        nan |                     nan |
Printing complete results for all models
| Category              | Benchmark   | Subtask   |   Accuracy | Number few shot   | Model       |
|:----------------------|:------------|:----------|-----------:|:------------------|:------------|
| commonsense_reasoning | copa        |           |       0.54 | 0-shot            | mpt-125m-hf |


Generating responses...
####################################################################################################
The answer to life, the universe, and happiness is the a- the the the the the**](#. the the a,,s a the the it the the,, and a the in the the the the the the to,, the the of is and the is of the the the х the,,,, the past the the in is Stanton,,,, is the with to, a a disclosure,,, the is, the the is is,, the,, the,, and, the the the holiday the, a. theoque the the is the,, the the,, the to, the a,s the theoquecpy the a the the the,ocols login the the
 of, and a,, a the the 01,, in,, to in the the the the thecircuit the,, a,,,, of the the,,,, a the the the the the is holiday the Hoffised, Scots,,, a, thecircuit the
,,,,,, the andExpect the past,, of a the the the the a the the, the-,,’ocols.., the the the the the the,, the the the the the.,, in and theunds,, the.,,, the is
####################################################################################################
Here's a quick recipe for baking chocolate chip cookies: Start by, in the the, and and the the, of the the, and the the to, the the to the the the the is, the, in the theundsevent and and, is aWe the the is the and the the the to, the the the,, the the the the, you a the,, the the the the the, the,,, in., the, the the, the the the’ the the and the the,, to the the the the the theALTH the the, the. the. and,, the the the the is the in to, the the the the the and and,, the,,, the the past’circuit,,,, in the the the the holiday the the, arien to a the the,, the the Jake the the the the, the the, theWe the and the the the, the the the the the the the in, the. a the login past,,, the a the the the the the, of the the of, the the the the in the, Engine, the,,,,We the the,, the the the the the,’ is the,, ofs the the. the, the,’, the and the and
####################################################################################################
bs=2, input_tokens=array([11, 13]), output_tokens=array([256, 256])
total_input_tokens=24, total_output_tokens=512
encode_latency=2.07ms, gen_latency=4134.89ms, decode_latency=0.69ms, total_latency=4137.66ms
latency_per_output_token=8.08ms/tok
output_tok_per_sec=123.74tok/sec
```

**Comment by vishalbakshi** [comment]
Closing this in lieu of #23.
